File: ios/StakDemo/Onboarding/TasteModel.swift
import SwiftUI

/// One taste bar — label, strength word (its own gray), fill fraction of the track.
struct TasteBar: Identifiable {
	let label: String
	let strength: String
	let strengthColor: Color
	let fraction: CGFloat
	var id: String { label }
}

/// The onboarding answers, turned into the taste reveal, the risk style and the
/// Profile chips (product audit, 2026-09-05: the reveal used to be fixed copy and the
/// quiz answers went nowhere). Deterministic and tiny: each taste is the share of the
/// user's brand picks that belong to it, nudged by the goal and the risk answer.
/// Mirrors android TasteModel.kt.
enum TasteModel {
	private static let tech: Set<String> = ["Apple", "Tesla", "Spotify", "Netflix", "Amazon", "Microsoft", "NVIDIA", "PlayStation", "Coinbase", "Uber"]
	private static let growth: Set<String> = ["Tesla", "NVIDIA", "Coinbase", "Uber", "Amazon", "Spotify", "Netflix"]
	private static let consumer: Set<String> = ["Nike", "Disney", "Netflix", "Spotify", "Amazon", "PlayStation", "Uber", "Apple", "Tesla"]
	private static let income: Set<String> = ["Apple", "Microsoft", "Nike", "Disney"]

	// 04 Goal options, in card order.
	static let goalLearn = 0, goalGrow = 1, goalFirstStocks = 2, goalExplore = 3
	// 05 Risk options, in card order.
	static let riskBuyMore = 0, riskHold = 1, riskStepAway = 2, riskSellSome = 3

	static func riskStyle(_ risk: Int) -> String {
		switch risk {
		case riskBuyMore: return "Growth-Oriented"
		case riskHold: return "Balanced"
		case riskStepAway: return "Conservative"
		case riskSellSome: return "Cautious"
		default: return "Growth-Oriented"
		}
	}

	/// tech / growth / consumer / income, each 0...1.
	static func scores(_ picks: Set<String>, goal: Int, risk: Int) -> [CGFloat] {
		func share(_ group: Set<String>) -> CGFloat {
			picks.isEmpty ? 0 : CGFloat(picks.filter { group.contains($0) }.count) / CGFloat(picks.count)
		}
		let t = share(tech) + (goal == goalFirstStocks ? 0.1 : 0)
		let g = share(growth) + (goal == goalGrow ? 0.25 : 0) + (risk == riskBuyMore ? 0.2 : 0) - (risk == riskSellSome ? 0.2 : 0)
		let c = share(consumer)
		let i = share(income) + (risk == riskSellSome ? 0.3 : 0) + (risk == riskStepAway ? 0.15 : 0) + (goal == goalLearn ? 0.1 : 0)
		return [t, g, c, i].map { min(max($0, 0), 1) }
	}

	static func bars(_ picks: Set<String>, goal: Int, risk: Int) -> [TasteBar] {
		let labels = ["Tech curious", "Growth seeking", "Consumer brands", "Income & dividends"]
		return zip(labels, scores(picks, goal: goal, risk: risk)).map { label, s in
			let strength = strengthOf(s)
			return TasteBar(label: label, strength: strength, strengthColor: colorOf(strength), fraction: 0.2 + 0.8 * s)
		}
	}

	/// The Profile's YOUR TASTE chips, strongest first, at most three.
	static func chips(_ picks: Set<String>, goal: Int, risk: Int) -> [String] {
		let s = scores(picks, goal: goal, risk: risk)
		var out: [(CGFloat, String)] = []
		if s[0] >= 0.3 { out.append((s[0], "Tech Curious")) }
		if s[1] >= 0.45 { out.append((s[1], "High Growth")) }
		if s[2] >= 0.3 { out.append((s[2], "Consumer Brands")) }
		if s[3] >= 0.35 { out.append((s[3], "Income Focused")) }
		if risk == riskSellSome || risk == riskStepAway { out.append((0.5, "Risk Aware")) }
		if goal == goalLearn { out.append((0.4, "Learning")) }
		var seen = Set<String>()
		let chips = out.sorted { $0.0 > $1.0 }.map { $0.1 }.filter { seen.insert($0).inserted }.prefix(3)
		return chips.isEmpty ? ["Just Exploring"] : Array(chips)
	}

	private static func strengthOf(_ s: CGFloat) -> String {
		if s >= 0.6 { return "Strong" }
		if s >= 0.3 { return "Medium" }
		return "Light"
	}

	// The authored strength colours (1:687): teal / muted / faint.
	private static func colorOf(_ strength: String) -> Color {
		switch strength {
		case "Strong": return Color(argb: 0xFF69B3CA)
		case "Medium": return Color(argb: 0xFF819ABB)
		default: return Color(argb: 0xFF5C6B85)
		}
	}
}

File: android/app/src/main/java/com/stak/demo/ui/onboarding/TasteModel.kt
package com.stak.demo.ui.onboarding

import com.stak.demo.ui.theme.stakColor
import androidx.compose.ui.graphics.Color

/**
 * The onboarding answers, turned into the taste reveal, the risk style
 * and the Profile chips (product audit, 2026-09-05: the reveal used to be
 * fixed copy and the quiz answers went nowhere). Deterministic and tiny:
 * each taste is the share of the user's brand picks that belong to it,
 * nudged by the goal and the risk answer. Mirrors ios TasteModel.swift.
 */
internal object TasteModel {
	private val TECH = setOf("Apple", "Tesla", "Spotify", "Netflix", "Amazon", "Microsoft", "NVIDIA", "PlayStation", "Coinbase", "Uber")
	private val GROWTH = setOf("Tesla", "NVIDIA", "Coinbase", "Uber", "Amazon", "Spotify", "Netflix")
	private val CONSUMER = setOf("Nike", "Disney", "Netflix", "Spotify", "Amazon", "PlayStation", "Uber", "Apple", "Tesla")
	private val INCOME = setOf("Apple", "Microsoft", "Nike", "Disney")

	// 04 Goal options, in card order.
	const val GOAL_LEARN = 0
	const val GOAL_GROW = 1
	const val GOAL_FIRST_STOCKS = 2
	const val GOAL_EXPLORE = 3

	// 05 Risk options, in card order.
	const val RISK_BUY_MORE = 0
	const val RISK_HOLD = 1
	const val RISK_STEP_AWAY = 2
	const val RISK_SELL_SOME = 3

	/** One reveal bar: label, strength word, its colour, fill fraction of the track. */
	data class TasteBar(val label: String, val strength: String, val strengthColor: Color, val fraction: Float)

	fun riskStyle(risk: Int): String = when (risk) {
		RISK_BUY_MORE -> "Growth-Oriented"
		RISK_HOLD -> "Balanced"
		RISK_STEP_AWAY -> "Conservative"
		RISK_SELL_SOME -> "Cautious"
		else -> "Growth-Oriented"
	}

	/** tech / growth / consumer / income, each 0..1. */
	fun scores(picks: Set<String>, goal: Int, risk: Int): List<Float> {
		fun share(group: Set<String>): Float = if (picks.isEmpty()) 0f else picks.count { it in group }.toFloat() / picks.size
		val tech = share(TECH) + (if (goal == GOAL_FIRST_STOCKS) 0.1f else 0f)
		val growth = share(GROWTH) + (if (goal == GOAL_GROW) 0.25f else 0f) + (if (risk == RISK_BUY_MORE) 0.2f else 0f) - (if (risk == RISK_SELL_SOME) 0.2f else 0f)
		val consumer = share(CONSUMER)
		val income = share(INCOME) + (if (risk == RISK_SELL_SOME) 0.3f else 0f) + (if (risk == RISK_STEP_AWAY) 0.15f else 0f) + (if (goal == GOAL_LEARN) 0.1f else 0f)
		return listOf(tech, growth, consumer, income).map { it.coerceIn(0f, 1f) }
	}

	fun bars(picks: Set<String>, goal: Int, risk: Int): List<TasteBar> {
		val labels = listOf("Tech curious", "Growth seeking", "Consumer brands", "Income & dividends")
		return labels.zip(scores(picks, goal, risk)).map { (label, s) ->
			val strength = strengthOf(s)
			TasteBar(label, strength, colorOf(strength), 0.2f + 0.8f * s)
		}
	}

	/** The Profile's YOUR TASTE chips, strongest first, at most three. */
	fun chips(picks: Set<String>, goal: Int, risk: Int): List<String> {
		val (tech, growth, consumer, income) = scores(picks, goal, risk)
		val out = mutableListOf<Pair<Float, String>>()
		if (tech >= 0.3f) out += tech to "Tech Curious"
		if (growth >= 0.45f) out += growth to "High Growth"
		if (consumer >= 0.3f) out += consumer to "Consumer Brands"
		if (income >= 0.35f) out += income to "Income Focused"
		if (risk == RISK_SELL_SOME || risk == RISK_STEP_AWAY) out += 0.5f to "Risk Aware"
		if (goal == GOAL_LEARN) out += 0.4f to "Learning"
		val chips = out.sortedByDescending { it.first }.map { it.second }.distinct().take(3)
		return chips.ifEmpty { listOf("Just Exploring") }
	}

	private fun strengthOf(s: Float): String = when {
		s >= 0.6f -> "Strong"
		s >= 0.3f -> "Medium"
		else -> "Light"
	}

	// The authored strength colours (1:687): teal / muted / faint.
	private fun colorOf(strength: String): Color = when (strength) {
		"Strong" -> stakColor(0xFF69B3CA)
		"Medium" -> stakColor(0xFF819ABB)
		else -> stakColor(0xFF5C6B85)
	}
}
